src/flatten_to_stream.rs

```diff
@@ -0,0 +1,101 @@
+use core::mem;
+use core::marker;
+use {Async, Future, Poll};
+use stream::Stream;
+
+/// Future for the `flatten_to_stream` combinator, flattening a future-of-a-stream to get just
+/// the result of the final future as a stream.
+///
+/// This is created by the `Future::flatten_to_stream` method.
+#[must_use = "streams do nothing unless polled"]
+pub struct FlattenToStream<F>
+    where
+        F : Future,
+        <F as Future>::Item: Stream<Error=F::Error>,
+{
+    state: FlattenToStreamState<F>
+}
+
+pub fn new<F>(f: F) -> FlattenToStream<F>
+    where
+        F : Future,
+        <F as Future>::Item: Stream<Error=F::Error>,
+{
+    FlattenToStream {
+        state: FlattenToStreamState::Future(f)
+    }
+}
+
+/// Stream that emits single error
+enum FailedState<T, E> {
```

Comment:
> Does this have to have `T` as a type parameter?

To implement `Stream` for `FailedState`, yes.

It is not strictly necessary to implement `Stream` (`T` could be a type parameter of `poll` function), but it makes almost no difference in number of lines of code.

> Also, could this be folded into the state below to avoid nested enums?

Technically, yes, but it would significantly compilcate a loop in `FlattenToStream`:

```
fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
    loop {
        self.state = match self.state {
            ...
            State::Stream(ref mut s) => {
                // This branch relies on the fact that
                // `self.state` remains unchanged after return.
                // So I can't (easily) do `mem::replace(&mut self.state, ..)`
                // with some dummy value.
                return s.poll();
            }
            State::Error(e) => {
                // This branch needs to take ownership of `e` to return it
                // but rust does not allow one match branch to take a reference
                // and other branch to take a value.
                ...
            }
        };
    }
```